### PR TITLE
LanguagePatcher Dictionary Error.

### PR DESCRIPTION
**Pull request by SamuramongeDev**
### Changes made in this pull request

  - Fixed an error where it tried to access an item from an empty dictionary. (`if (_customLines is null || _customLines.Count == 0) return;`)

### CODE REVIEW

**Comment by JKohlman** [comment]
I'm a little confused as to your goal here for two reasons:

(1) The dictionary can't be null unless something went terrible wrong as it's a static member
```csharp
private static readonly Dictionary<string, Dictionary<string, string>> _customLines = new();
```
(2) How does checking if the dictionary has any arbitrary item in it prevent errors?

If your goal is to just ensure that lines 48 & 49 don't throw an error I'd suggest instead checking to see if the specific objects needed for the function exist in the dictionary (`.has(...)`) or use `TryGetValue` to return a boolean and also get the value in the same line.

**Comment by SamuramongeDev** [comment]
> I'm a little confused as to your goal here for two reasons:
> 
> (1) The dictionary can't be null unless something went terrible wrong as it's a static readonly member
> 
> ```cs
> private static readonly Dictionary<string, Dictionary<string, string>> _customLines = new();
> ```
> 
> (2) How does checking if the dictionary has any arbitrary item in it prevent errors?
> 
> If your goal is to just ensure that lines 48 & 49 don't throw an error I'd suggest instead checking to see if the specific objects needed for the function exist in the dictionary (`.has(...)`) or use `TryGetValue` (see line 34) to return a boolean and also get the value in the same line.

When i first launched the game, i saw an error caused by LanguagePatcher.cs trying to acces an item on that dictionary. So, by common sense, i added a check there, and it worked!
Currently checking the code.

**Comment by JKohlman** [comment]
> When i first launched the game, i saw an error caused by LanguagePatcher.cs trying to acces an item on that dictionary. So, by common sense, i added a check there, and it worked! Currently checking the code.

In that case I'd suggest using one of the two methods I described instead as they _should_ work and would be a little more accurate to the code.

**Comment by LeeTwentyThree** [comment]
Exiting early without any sort of message/error is never a preferred solution.
However I do see potential for error here: https://github.com/SubnauticaModding/Nautilus/blob/aa6a3dffba70159b805d73e222679fc28ad5133a/Nautilus/Patchers/LanguagePatcher.cs#L52-L53

**Comment by SamuramongeDev** [comment]
returning that early from the method is necessary,
```
var fallbackStrings = _customLines[FallbackLanguage];
var currentStrings = _customLines[_currentLanguage];
```
those rely on `_customLines` whitch it's empty in this case. So the solution would be checking if `_customLines ` it's empty or not.

**Comment by SamuramongeDev** [comment]
> Exiting early without any sort of message/error is never a preferred solution. However I do see potential for error here:
> 
> https://github.com/SubnauticaModding/Nautilus/blob/aa6a3dffba70159b805d73e222679fc28ad5133a/Nautilus/Patchers/LanguagePatcher.cs#L52-L53

Yeah, so, what we can do in that case? i tested and those cause errors.

**Comment by LeeTwentyThree** [comment]
The issue is likely rooted elsewhere, and returning early is only a superficial fix.

**Comment by JKohlman** [comment]
> > Exiting early without any sort of message/error is never a preferred solution. However I do see potential for error here:
> > https://github.com/SubnauticaModding/Nautilus/blob/aa6a3dffba70159b805d73e222679fc28ad5133a/Nautilus/Patchers/LanguagePatcher.cs#L52-L53
> 
> Yeah, so, what we can do in that case? i tested and those cause errors.

Check out how it's done elsewhere in the file, particularly either lines 34 & 35 for the `TryGetValue` method or lines 94-98 for the `Contains` method

**Comment by LeeTwentyThree** [comment]
I can't even figure out what's going on tbh, docs would be much appreciated. @Metious wrote this I believe, so we can wait on him if needed.

**Review by Metious — CHANGES_REQUESTED** [review]
Based on my tests, you will get an NRE when no mod has registered any language entries, and you will also get an error if there are language entries but a specific language does not have any entries.

Your changes fix the former error, but the latter is still persistent.

Like mentioned in the other comments, since the `_customLines` dictionary is never null, you can change the first check to the following:
```csharp
if (!_customLines.TryGetValue(FallbackLanguage, out var fallbackStrings) & !_customLines.TryGetValue(_currentLanguage, out var currentStrings))
{
    return;
}
```
With that change, you will no longer need line 52-53. https://github.com/SubnauticaModding/Nautilus/blob/aa6a3dffba70159b805d73e222679fc28ad5133a/Nautilus/Patchers/LanguagePatcher.cs#L52-L53


For the second issue, there are a lot of locations for possible error, but all can be fixed by using the [null-coalescing assignment operator](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/null-coalescing-operator) for `fallbackStrings` and `currentStrings` dictionaries after the check. With these changes, the end result should look like this:
```csharp
if (!_customLines.TryGetValue(FallbackLanguage, out var fallbackStrings) & !_customLines.TryGetValue(_currentLanguage, out var currentStrings))
{
    return;
}

fallbackStrings ??= new();
currentStrings ??= new();
```

Please note that the logical AND operator (`&`)  is essential for the early check because both of the `TryGetValue` calls must be called even if the first check returns `false`. 

For more information on the difference between logical AND operator (`&`) and the conditional logical AND operator (`&&`), click [here](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/operators/boolean-logical-operators#logical-and-operator-).

File: Nautilus/Patchers/LanguagePatcher.cs [hunk 1]
```diff
@@ -1,112 +1,117 @@
...
+
+    internal static void InsertCustomLines(ref Language __instance)
+    {
+        if (!_customLines.TryGetValue(FallbackLanguage, out var fallbackStrings) & !_customLines.TryGetValue(_currentLanguage, out var currentStrings))
+        {
+          return;
+        }
+
+        fallbackStrings ??= new();
+        currentStrings ??= new();
```
Comment (SamuramongeDev) [inline Nautilus/Patchers/LanguagePatcher.cs:54]: This is the change.

**Pull request closed without merging**